### DISCUSSION TITLE
skynet添加更高精度timer的问题

### DISCUSSION THREAD

**Discussion by Dllinsanity** [cloudwu/skynet#1916] [Q&A] [answered]
实现和视景程序通讯服务时，需要更稳定的发送间隔，使用skynet自带的timeout周期波动对视景显示效果有影响。为了快速解决这个问题，我使用c实现的这个服务，并且在服务里创建线程计算时钟提供更准确的调度。由于这个方案路子太野，后续我在框架里实现hrtimer并提供和已有timer一样的使用方式，但是由于cb函数被调用时机的问题，这个高精度定时器还是会有波动。那么我只能魔改框架提供另外一个callback函数给服务用以注册，以支持更高精度更稳定的timer。我想请问，有没有更优雅的方案完成我的需求，谢谢各位大佬。

**Comment by sniper00** [comment]
非实时系统都会有波动，你可以添加 fix 补偿机制

**Reply by Dllinsanity** [reply]
谢谢回复，我第一版解决方案是在处理timeout的时候补偿的，能保证误差不被叠加，但是视景看起来还是会抖动。现在我想在skynet核心库添加一个模块，在这个模块里补偿，每个服务向这个模块注册callback。这是这个模块在windows下的一个简单实现https://github.com/Dllinsanity/libschedule 。只是我觉得这样做破坏了skynet本来的设计，不够优雅。

**Comment by cloudwu** [comment, marked as answer]
我觉得非侵入的方法是你另外在本地启一个进程，通过本地按周期发 udp 包。skynet 应该可以保证及时处理 udp 消息。
